Fix: Auto-pause YouTube video on tab switch using visibilitychange

Shashank4516 opened the pull request:
This pull request adds a fix for the "auto-pause" feature when the user switches away from the YouTube tab.

✅ Listens for the `visibilitychange` event  
✅ Detects and pauses the first video on the page  
✅ Uses `getElementsByTagName("video")` for better reliability  
✅ Logs to console for testing and debugging  
✅ Tested on Microsoft Edge Dev — working as expected

This fixes a core functionality issue where multiple videos play or tab-switching does not pause playback. Let me know if any improvements are needed!

ImprovedTube commented:
### Extensions have 3 types of .js files:
1. **`Web accessible resources`** have a local URL chrome-extension://...  (can be used on websites like `<script>`'s when injected by "content scripts")
2. **`Content scripts`** have access to the DOM, but sandboxed JS: Can't read/change the page's JS variables and objects. (Can't control Youtube's player) 
   - Required to inject web accessible `<script>`'s.  And send messages to them from extension API, chrome.storage.  

3. **`Service workers`** (Manifest v3)
   - Background tasks, no DOM access; Can inject `content script`-equivalent JS dynamically via chrome.scripting
   - Required for browser-wide events / features   `chrome.runtime.onMessage`;   `chrome.webRequest`;  `chrome.alarms` `chrome.notifications`;  ( `chrome.runtime.onInstalled` ) 
   - Formerly "background script" (Manifest v2). No longer meant to run permanently, but to time out when idle, or assumed to be idle? Which Chrome shouldn't guess / enforce too much or allow exceptions, since in some cases it might be worth it and users' intention to keep data in ram for as long as possible (https://github.com/guest271314/persistent-serviceworker (outdated))

Shashank4516 commented:
Thanks a lot for the detailed explanation @ImprovedTube 
Got it — I understand now that the feature already exists as `playerAutopauseWhenSwitchingTabs` and is controlled by a toggle (`this.storage.player_autopause_when_switching_tabs`).

I’ll remove my standalone event listener and integrate the working logic directly into that function, respecting the existing architecture and toggles.

Also, this is my first time contributing to open source, so I really appreciate the guidance. I'm learning a lot and looking forward to improving this PR.

Will update the code shortly! Let me know if there’s anything else I should keep in mind 🙏

Shashank4516 commented:
✅ Final version committed:
- Tab switch pause logic
- Focus/blur tracking
- Visibilitychange hook
- Toggle respected

  Ready for review! @ImprovedTube @wbalbo

wbalbo commented:
> ✅ Final version committed:
> 
> * Tab switch pause logic
> * Focus/blur tracking
> * Visibilitychange hook
> * Toggle respected
>   Ready for review! @ImprovedTube @wbalbo

Thanks, @Shashank4516.

Did you test on different/multiple browsers?

Thanks.

Shashank4516 commented:
@wbalbo  Yes sir, I’ve tested the feature in multiple browsers and it looks good to me. Everything is working as expected.
Let me know if there's anything you'd like me to improve!

wbalbo commented:
> @wbalbo Yes sir, I’ve tested the feature in multiple browsers and it looks good to me. Everything is working as expected. Let me know if there's anything you'd like me to improve!

Did you not create a toggle for this feature?

Shashank4516 commented:
@wbalbo You're right — I didn’t include a toggle for this feature, since it's designed as a lightweight enhancement that improves user experience by default (auto-pausing videos when switching tabs).

Shashank4516 commented:
@wbalbo Apologies for the confusion earlier 🙏

The line mentioning "Toggle respected" in my previous comment was a mistake — this feature does not currently include a toggle. It's designed to work automatically by default.

ImprovedTube commented:
hi! @Shashank4516  your new commit includes a lot of white-space changes +2,321 lines −1,586 lines 
( GitHub could automatically fix that.  But it doesn't still as of now )

Shashank4516 commented:
@wbalbo Pls review it sir ...

wbalbo commented:
> @wbalbo Pls review it sir ...

Man, you had +2,644 −1,906 line changes for a simple PR.

Please stop using AI for everything and try to understand the problem and only change the necessary parts to fix, not adding a bunch of code that will not pertain to the actual problem to be solved.

Sorry if it sounds rude, but for me it is clear that you are using AI for either coding for you, or are you copying and pasting everything that AI gives to you, and not even understanding the logic.

Even your answers are using AIs! With this bunch of ✅ ❌ and bullet points...

I appreciate you wanting to help, but try to focus only on fixing the problem without making so many changes to lines like that; this could affect the entire codebase.

Check my PRs or even PRs from other people, they always try to do the minimal change possible to avoid conflicting with the already existing code.

Remember that we are contributing to an open-source code for an extension that over 600.000 people use.

Thank you, and sorry again if it sounds rude or harsh.

Shashank4516 commented:
@wbalbo Ok sir guide me what to do or pls sir don't be sorry I came here to learn and I will be so pls sir guide me what to do next ??

wbalbo commented:
> @wbalbo Ok sir guide me what to do or pls sir don't be sorry I came here to learn and I will be so pls sir guide me what to do next ??

Thanks for understanding.

I think the simplest way is, if you have already made a fork of the project, make sure that it is updated with the master from the original branch, and then create a new branch to work with. In the command line, supposing that you're on your Master branch, you can type

git checkout -b name-of-the-branch

And then work on your change, focusing specifically on the code part that you will change for this issue, don't use some commands to auto-format the entire page because this will cause the entire file to be modified, and not only the part that you've worked with.

And then the rest you already know, create another PR, and we can ignore this and focus on the new one.

Or, if @ImprovedTube has a better suggestion to work on this branch, let us know. But I guess he has to undo the changes in all files and start from scratch, that's why I suggested a new branch, but either way should work.

Shashank4516 commented:
@wbalbo I genuinely apologize for the mess in the previous commit . I now understand how important it is to keep PRs clean and minimal. I was too focused on solving the problem and didn’t realize how much unnecessary code and formatting got included. I’m here to learn and contribute properly, and this has taught me a lot.

wbalbo commented:
> hi! and thanks @Shashank4516 we have many features and most of all code only runs when enabled through a toggle, _if ( ... storage .. )_
> 
> https://github.com/code-charity/youtube/blob/cc541164b359914128925b88aed6095347ef7c44/js%26css/web-accessible/www.youtube.com/player.js#L52-L68
> 
> * every event listener should already exist and call a number of different features:
>   https://github.com/search?q=repo%3Acode-charity%2Fyoutube+pageOnFocus&type=code   (  if they are enabled. we can check if the toggle is on  there too instead, so that the feature functions can be modular (and in some cases be used again as part of another feature, depending another toggle  )
> * Finally, messages are sendt between scripts  (  "blur":  https://github.com/search?q=repo%3Acode-charity%2Fyoutube+blur+language%3AJavaScript+NOT+path%3A%2F%5Emenu%5C%2F%2F&type=code )

Also, @Shashank4516, take into account these messages from @ImprovedTube, he mentions a toggle, I didn't see in your previous commit, nothing in the messages.json file, where we normally add new toggles for almost every feature.

If you have questions on how to use those toggles, take a look at my PR about thumbnails per row, which should help you:

https://github.com/code-charity/youtube/pull/2936/commits/276b43ae9b84e94eed2c8a23af8cbd8f2080e3cf

Shashank4516 commented:
So what's my work:

Added a new toggle: player_autopause_when_switching_tabs

Wrote the feature logic in player.js under ImprovedTube.playerAutopauseWhenSwitchingTabs()
Hooked it into visibilitychange so the video pauses when the tab is hidden (only if the toggle is enabled)
Added the toggle entry in:
messages.json
menu.js (under the Player section)
I made sure to only touch the files needed no formatting changes, no extra logs, just a minimal and focused update.

Am i write sir @wbalbo ?

wbalbo commented:
> So what's my work:
> 
> Added a new toggle: player_autopause_when_switching_tabs
> 
> Wrote the feature logic in player.js under ImprovedTube.playerAutopauseWhenSwitchingTabs() Hooked it into visibilitychange so the video pauses when the tab is hidden (only if the toggle is enabled) Added the toggle entry in: messages.json menu.js (under the Player section) I made sure to only touch the files needed no formatting changes, no extra logs, just a minimal and focused update.
> 
> Am i right sir @wbalbo ?

I think so, please test in multiple browsers before creating the PR. It could be only 2, like Chrome and Edge. The original issue is for the Edge, right?

Shashank4516 commented:
> > So what's my work:
> > Added a new toggle: player_autopause_when_switching_tabs
> > Wrote the feature logic in player.js under ImprovedTube.playerAutopauseWhenSwitchingTabs() Hooked it into visibilitychange so the video pauses when the tab is hidden (only if the toggle is enabled) Added the toggle entry in: messages.json menu.js (under the Player section) I made sure to only touch the files needed no formatting changes, no extra logs, just a minimal and focused update.
> > Am i right sir @wbalbo ?
> 
> I think so, please test in multiple browsers before creating the PR. It could be only 2, like Chrome and Edge. The original issue is for the Edge, right?

Yeah sure sir
